cpu: LoopPoint analysis object

studyztp opened the pull request:
Adding SimObject and ProbeListener objects for gem5 to perform LoopPoint analysis.
The original paper for the LoopPoint paper: https://ieeexplore.ieee.org/document/9773236
LoopPoint is a sampling methodology that uses the program counter of a set of selected loops and their number of iteration to mark the execution points in a macro-benchmark for the defined regions.

Therefore, the LoopPont analysis object in gem5 will 
1. Define regions with the number of user-level instruction executed
2. Collect loop iterations for every region
3. Collect user-level basic block vector for every region
4. Exclude specific memory ranges to avoid analyzing multi-threaded libraries
5. Raise an exit event at the end of every region

The LoopPoint analysis object in gem5 is only tested with full-system simulation.

studyztp committed:
cpu: add basic files for LoopPoint analysis

Because the LoopPoint analysis will be done with ATOMIC CPU, so all
files related to the LoopPoint analysis object will be under
/src/cpu/simple/probes.

Change-Id: Icbdb0742b712a23dc8f6a19f4c1c827a1f5bf288

studyztp committed:
cpu: add looppoint_analysis.hh content and licenses

In looppoint_analysis.hh, added LooppointAnalysis and
LooppointAnalysisManager classes.
Added all functions and variables for the classes.
Comments needed.

Change-Id: Ia7425b672ef092a68c99b702136850bfa1fcf0a2

studyztp committed:
cpu: add comments and improve naming in looppoint_analysis.hh

Add comments to most variables and functions.
Change the naming of some variables and functions to improve the
clearness.

Change-Id: Idb557ec84698b4344ed4683f5de87b1a3c2fd66d

studyztp committed:
cpu: add looppoint_analysis.cc content

Add LooppointAnalysis and LooppointAnalysisManager function definitions

Change-Id: I1c05072ebf1b744ee102a82f8de2b93bab4a056f

studyztp committed:
cpu: add the python class

Add the python classes for the LooppointAnalysis and the
LooppointAnalysis Manager.

Change-Id: I0a882bc1a9ef03b7b482e871a7160e7c33f9ac08

studyztp committed:
cpu: fix a wrong file path

Change-Id: I93343f4053c7a6d1bd4b6972a1e7c3dbc073c979

studyztp committed:
cpu: fix a issue

Change-Id: Iab621e294c84c7f5c704882b0c681f950ad08f9c

studyztp committed:
cpu: add more comments to looppoint_analysis.cc

Change-Id: I027db66ffed0cd5957bae2a9a36286ca1c73c313

studyztp committed:
cpu: add more debug flags

Change-Id: I4edd8f383294f76d3e76895d3a631cba21a45f90

powerjg commented on src/cpu/simple/probes/LooppointAnalysis.py:
```diff
@@ -0,0 +1,90 @@
+        PyBindMethod("getMostRecentBackwardBranchCount"),
+    ]
+
+    region_length = Param.Int(100000000, "the length of the region")
```
Can you add a pointer to the loop point paper here? Use the `add_citation` function, IIRC.

You may also want to write up a short "paper" on how to use loop point and sim point in gem5, put it on arxiv and add that as a citation. You've done lots of work here, and you should get some credit :)

studyztp replied:
I added the citation with the `add_citation` function. 

Thanks! I'll prepare a write-up on how to use them :)

powerjg commented on src/cpu/simple/probes/LooppointAnalysis.py:
```diff
@@ -0,0 +1,90 @@
+    )
+    marker_valid_addr_range = Param.AddrRange(
+        AddrRange(start=0, end=0), "the valid address range for markers"
+    )
```
Does it make sense to have defaults here?

studyztp replied:
You are right. I removed the default and hint on inputting `AddrRange(start=0, end=0)` in the parameter descriptions.

powerjg commented on src/cpu/simple/probes/LooppointAnalysis.py:
```diff
@@ -0,0 +1,90 @@
+        [], "the excluded address ranges for basic blocks"
+    )
+    if_listening = Param.Bool(
+        True, "if the LooppointAnalysis is listening to " "the probe point"
```
```suggestion
        True, "if the LooppointAnalysis is listening to the probe point"
```

powerjg commented on src/cpu/simple/probes/LooppointAnalysis.py:
```diff
@@ -0,0 +1,90 @@
+        PyBindMethod("getMostRecentBackwardBranchCount"),
+    ]
+
+    region_length = Param.Int(100000000, "the length of the region")
```
```suggestion
    region_length = Param.Int(100_000_000, "the length of the region")
```

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+                            "simulation? %s\n", ifListening ? "Yes" : "No");
+
+    for (int i = 0; i < params.bb_excluded_addr_ranges.size(); i++)
+    {
```
```suggestion
    for (int i = 0; i < params.bb_excluded_addr_ranges.size(); i++) {
```

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+    else
+    {
+        localBBV.find(pc)->second++;
+    }
```
I'm not going to do this for the whole file, but please change this style

```suggestion
    if (localBBV.find(pc) == localBBV.end()) {
        localBBV.insert(std::make_pair(pc, 1));
    } else {
        localBBV.find(pc)->second++;
    }
```

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+
+void
+LooppointAnalysis::checkPc(const std::pair<SimpleThread*,
+                                                     StaticInstPtr>& inst_pair)
```
Align with the `(`

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+    {
+        for (int i = 0; i < bbExcludedAddrRanges.size(); i++)
+        {
+            if (pc >= bbExcludedAddrRanges[i].start() &&
```
Can you use `contains` (https://github.com/gem5/gem5/blob/stable/src/base/addr_range.hh#L489)?

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+        lpaManager->updateGlobalBBV(pc);
+        bbInstCounter = 0;
+
+        if (markerValidAddrRange.end() > 0 &&
+         (pc < markerValidAddrRange.start() || pc > markerValidAddrRange.end())
```
`contains`?

powerjg commented on src/cpu/simple/probes/looppoint_analysis.cc:
```diff
@@ -0,0 +1,280 @@
+    if (ifListening)
+    {
+        if (listeners.size() == 0) {
+            listeners.push_back(new looppointAnalysisListener(this,
```
See #1439 to make sure this doesn't conflict

studyztp replied:
It does conflict because I am using the constructor of the ProbeListener object to add the listeners to the probe point and remove them using its destructor too; but I think I can push a patch to fix it later after we decide what is the best way to add the listeners in that pull request.

powerjg commented on src/cpu/simple/probes/looppoint_analysis.hh:
```diff
@@ -0,0 +1,331 @@
+     * This set stores the Program Counter addresses of the valid backward
+     * branches.
+     */
+    std::unordered_set<Addr> backwardBranchPC;
```
Is there a reason you can't use `backwardBranchCounter` for this?

studyztp replied:
It is a great catch! I removed `backwardBranchPC`and started using `backwardBranchCounter` instead now.

powerjg commented on src/cpu/simple/probes/looppoint_analysis.hh:
```diff
@@ -0,0 +1,331 @@
+        return mostRecentBackwardBranchPC;
+    };
+
+    std::unordered_map<Addr, uint64_t> getBackwardBranchCounter() const
```
We should be careful with this function. This map could be very large and this function will likely return a copy. Where is it called? Should it return a copy? should it return a reference?

studyztp replied:
It can be called at every end of the region. In the Python script, the map will be stored as a dictionary. We don't expect to change anything in the map. 
I think by default, Pybind converts it to dictionary that copied the map's data, so it doesn't return a reference, but since we don't expect to change anything in the map. returning a copy should be fine.

powerjg reviewed (commented):
Mostly small comments below. Great job!

studyztp committed:
cpu: modified after review feedback

src/cpu/simple/probes/LooppointAnalysis.py:
- remove default values for bb_valid_addr_range and
marker_valid_addr_range
- add more comments to explain parameter behaviors
- add citation to the LoopPoint paper

src/cpu/simple/probes/looppoint_analysis.cc:
- fix the incorrect styles
- remove updateBackwardBranch() function call
- match the style of checking if listeners vector is empty
- change the way of stopListening() to remove the listeners through the
manager instead of through the ProbeListener object's destructor.

src/cpu/simple/probes/looppoint_analysis.hh:
- removed backwardBranchPC and use the backwardBranchCounter to replace
its functionaility. Therefore, also removed updateBackwardBranch
function.

Change-Id: Id2430e2f04e61f72d5c4f1aad5cfd4d24a0fbc45

studyztp committed:
cpu: fix the incorrect debug message

Change-Id: I062e359e8c9205a9a993a33865434922c1f540b8

studyztp commented:
> Mostly small comments below. Great job!

Thank you for the detailed feedback! I addressed all of them and pushed the changes. I am really sorry for being late.

studyztp committed:
cpu: add function to get inst map of each basic block

Change-Id: I147d8c90cdfc7bf795d1c6a6daf96e11fa1c0858

studyztp committed:
cpu: make PcCountPair use 64 bit unsigned int for count

In PcCountPair param, change the type for "count" from 32 bit int to
64 bit unsigned int.

Change-Id: I2dc1bb2692914f06eaaae9bd5bbfb061bcbbfb8b

studyztp commented:
@BobbyRBruce Here is the PR for the LoopPoint Analysis. It is independent of the InstTracker. I noticed that this PR does not show up when I search for my username in the PRs.

BobbyRBruce commented on src/cpu/simple/probes/LooppointAnalysis.py:
```diff
@@ -0,0 +1,134 @@
+    bb_valid_addr_range = Param.AddrRange(
+        "the valid address range for basic blocks. If the address start and"
+        "end are both 0 (AddrRange(start=0, end=0)), it means every address is"
+        "valid."
```
When you have literal strings separated by whitespace (inc. newlines) in Python (and CPP too) they concatenated. So in this case the string passed to `Param.AddrRange`  will be "the valid address range for basic blocks. If the address start andend are both 0 (AddrRange(start=0, end=0)), it means every address isvalid." (note the lack of space between "and" and "end" and "is" and "valid").

Please add spaces where needed and in the other places in this PR this occurs.

studyztp replied:
Thanks for catching this. I fixed all places where this issue might occur.

BobbyRBruce reviewed (changes requested):
One tiny thing to fix, otherwise I'm fine with this and am happy to have it merged.

studyztp committed:
cpu: fix pc count pair helper function return type

Change the helper function's return type from int to uint64_t

Change-Id: I34b6b563a6333bbf8516a16d2ad4b76b7c16bfe4

studyztp committed:
cpu: fix looppoint anaylsis param python string spacing

Change-Id: I98fe434f1066f12b975425e49baca6e6a6087dab

studyztp commented:
> One tiny thing to fix, otherwise I'm fine with this and am happy to have it merged.

I fixed the params' python string spacing issue. I thin
